MSC1704: Adding ?via= to matrix.to permalinks to help with routing

turt2live opened the pull request:
[Rendered](https://github.com/matrix-org/matrix-doc/blob/travis/msc/matrix.to-permalinks/proposals/1704-matrix.to-permalinks.md)

KitsuneRal commented on proposals/1704-matrix.to-permalinks.md:
```diff
@@ -0,0 +1,36 @@
+highest power level in the room, provided that power level is 50 or higher. The other
+2 servers should be the most popular servers in the room based on the number of joined
+users. This same heuristic should apply to the first server if no user meets the power
+level requirements.
```
In the light of lazy-loading the user list, this implies the client has to force-load it? Also, eventually-to-be auditorium rooms are just not compatible with the logic. Can the list of servers (or even the entire permalink) be done on the server side?

turt2live replied:
Generating the permalink server side feels like too much of a stretch imo. Under this proposal, clients have some amount of freedom for how they pick servers. Clients could consider people who have talked recently for the server selection as they are also likely to stick around long enough for a permalink to work (if someone were to be handing a permalink to a friend). Similarly with lazy loading, you don't really need to force-load the entire member list: whatever members you're already aware of will probably be fine.

Another option could be to base the population check purely on the senders of various room state events (name, topic, power levels, etc) that aren't membership events. Could even extend this further to looking at the list of users in the power level event and prioritizing them higher.

KitsuneRal replied:
"People who talked recently being people who are likely to stick around long enough" is a bit of a stretch if the churn in the room is high. When the last 10 messages are "hi" followed by parting, I wouldn't expect these people's servers be a reliable routing.
I wonder if using aliases for the purpose of picking good-enough servers would be more reliable.

turt2live replied:
That was only one example. Clients aren't locked in to using one algorithm over another: just that they use one which gives them servers which are likely to be in the room in the future.

ara4n commented on proposals/1704-matrix.to-permalinks.md:
````diff
@@ -0,0 +1,36 @@
+https://matrix.to/#/!somewhere:example.org/$something:example.org?via=example-1.org&via=example-2.org
+```
+
+Clients can pass the servers directly to `/join` in the form of `server_name`
````
what are these `server_name` parameters of which you speak?

turt2live replied:
Query params of https://matrix.org/docs/spec/client_server/latest.html#post-matrix-client-r0-join-roomidoralias

ara4n replied:
ah, i misread /join as the slash-command rather than the API.

ara4n commented on proposals/1704-matrix.to-permalinks.md:
```diff
@@ -0,0 +1,36 @@
+parameters.
+
+When generating the permalinks, clients should pick servers that have a reasonably
+high chance of being in the room in the distant future. The current recommendation
```
should we try to use the server from the domain part of the room id?

turt2live replied:
imo servers should already be appending the server name in the room ID to the candidates so clients don't have to worry about it. There's no guarantee that the server name in the room ID is still a resident in the room, or has a chance of being around for a while so I'm further hesitant to recommend it.

turt2live commented:
There hasn't been a whole lot of feedback in the negative for this proposal, and I think I've addressed the concerns that were raised:

@mscbot fcp merge

non-Jedi commented on proposals/1704-matrix.to-permalinks.md:
````diff
@@ -0,0 +1,36 @@
+By adding a new parameter to the end, receivers can more easily join the room:
+
+```
+https://matrix.to/#/!somewhere:example.org?via=example-1.org&via=example-2.org
````
Note that this is a breaking change for any clients that parsed matrix.to urls using naive string-splitting. Per @turt2live in #matrix-spec:matrix.org, the clients he sampled didn't fall into this category and either handled matrix.to links like real URIs or didn't handle them at all. I don't see it as a huge issue (especially not compared to how awful including matrix.to in the spec at all is), but I thought it was worth documenting this concern on the PR.

turt2live commented:
Merged via https://github.com/matrix-org/matrix-doc/pull/1955 🎉
